src/lookup-table.c

```diff
@@ -0,0 +1,450 @@
+ *
+ * @param a The first hash value to compare.
+ * @param b The second hash value to compare.
+ * @return An integer indicating the result of the comparison.
+ */
+static inline int cmp_hash(const uint32_t a, const uint32_t b)
+{
+	// Compare hash values
+	if(a < b)
+		// First hash is less than second hash
+		return -1;
+	if(a > b)
+		// First hash is greater than second hash
+		return 1;
+
+	// Match
+	return 0;
+}
+
+/**
+ * @file datastructure.c
+ * @brief Implements a binary search algorithm for lookup_table structures.
+ *
+ * @param key Pointer to the lookup_table structure to search for (element to find).
+ * @param base Pointer to the base of the array of lookup_table structures to search in.
+ * @param nel Number of elements in the array to search in.
+ * @param try Pointer to a pointer where the address of the found element will be stored.
+ * @return true if the element is found, false otherwise.
+ *
+ * This function performs a binary search on an array of lookup_table structures.
+ * It compares the key with the middle element of the array and decides whether
+ * to search in the left or right half of the array. The search continues until
+ * the element is found or the array is exhausted. This function is used to find
+ * an element in a sorted array of lookup_table structures based on the hash value.
+ * Finding the element in the array is done in O(log n) time complexity.
+ *
+ * This function is different from the standard bsearch function in that it
+ * always the address of where the element would be found if it were in the
+ * array - even if the element is not found. This is useful for inserting new
+ * elements subsequently as we already have the correct position in the array
+ * without resorting to do a linear search to find the insertion point later on.
+ * We, instead, make the function return false if the element is not found.
+ *
+ * The standard library's bsearch() function, however, returns a pointer to the
+ * found element or NULL (!) if the element is not found which provides no
+ * information about the position where the element would need to be inserted.
+ */
+static bool binsearch(const struct lookup_table *base, const uint32_t hash, size_t nel, const struct lookup_table **try
+#ifdef MEASURE_ITERATIONS
+	, unsigned int *iterations
+#endif
+)
+{
+#ifdef MEASURE_ITERATIONS
+	// Count the number of iterations
+	*iterations = 0;
+#endif
+	// Run while there are elements left to be searched in the base array
+	while(nel > 0)
+	{
+#ifdef MEASURE_ITERATIONS
+		// Increment the number of iterations
+		(*iterations)++;
+#endif
+		// Set the try pointer to the (relative) middle element of the
+		// current base
+		*try = base + (nel / 2);
+
+		// Compare the key with the current element
+		const int sign = cmp_hash(hash, (*try)->hash);
+
+		if(sign == 0)
+		{
+			// If the key matches the element, we found at least one
+			// occurrence of the key in the array
+			return true;
+		}
+		else if(nel == 1)
```

Comment:
From a reading perspective I would have expected a check for `sign <0` here and only within this case a check for `nel == 1` and not the other way round

Reply:
but we can have `nel == 1` and `sign > 0` in which case we have the code
``` c
			// If the key is greater than the middle element, the key
			// would be inserted after the middle element
			(*try)++;
```